File: go/internal/httprequest/request_line.go
package httprequest

import "bytes"

// RequestLine is the course-defined representation of an HTTP/1.1 request
// line. Target and method-token grammar are intentionally deferred.
type RequestLine struct {
	Method  string
	Target  string
	Version string
	Valid   bool
}

// ParseRequestLine validates one CRLF-free request line.
func ParseRequestLine(line []byte) RequestLine {
	if bytes.ContainsAny(line, "\r\n\t") {
		return RequestLine{}
	}

	parts := bytes.Split(line, []byte(" "))
	if len(parts) != 3 || len(parts[0]) == 0 || len(parts[1]) == 0 || len(parts[2]) == 0 {
		return RequestLine{}
	}
	if !bytes.Equal(parts[2], []byte("HTTP/1.1")) {
		return RequestLine{}
	}

	return RequestLine{
		Method:  string(parts[0]),
		Target:  string(parts[1]),
		Version: string(parts[2]),
		Valid:   true,
	}
}

// String returns the shared-fixture representation of a parsed request line.
func (line RequestLine) String() string {
	if !line.Valid {
		return "invalid"
	}

	return line.Method + "|" + line.Target + "|" + line.Version
}

File: rust/src/request_line.rs
/// A course-defined HTTP/1.1 request line with owned byte fields.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RequestLine {
    pub(crate) method: Vec<u8>,
    pub(crate) target: Vec<u8>,
    pub(crate) version: Vec<u8>,
}

/// The request line did not match this course's limited HTTP/1.1 grammar.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RequestLineError {
    Invalid,
}

impl RequestLine {
    pub fn parse(line: &[u8]) -> Result<Self, RequestLineError> {
        if line.contains(&b'\r') || line.contains(&b'\n') || line.contains(&b'\t') {
            return Err(RequestLineError::Invalid);
        }

        let mut fields = line.split(|&byte| byte == b' ');
        let (Some(method), Some(target), Some(version)) =
            (fields.next(), fields.next(), fields.next())
        else {
            return Err(RequestLineError::Invalid);
        };

        if method.is_empty()
            || target.is_empty()
            || version != b"HTTP/1.1"
            || fields.next().is_some()
        {
            return Err(RequestLineError::Invalid);
        }

        Ok(Self {
            method: method.to_vec(),
            target: target.to_vec(),
            version: version.to_vec(),
        })
    }
}

impl std::fmt::Display for RequestLine {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            formatter,
            "{}|{}|{}",
            String::from_utf8_lossy(&self.method),
            String::from_utf8_lossy(&self.target),
            String::from_utf8_lossy(&self.version)
        )
    }
}

#[cfg(test)]
mod tests {
    use super::RequestLine;

    #[test]
    fn request_line_parses_three_single_space_fields() {
        let parsed = RequestLine::parse(b"GET /search?q=rust HTTP/1.1")
            .expect("request line should be valid");
        assert_eq!(parsed.method, b"GET");
        assert_eq!(parsed.target, b"/search?q=rust");
        assert_eq!(parsed.version, b"HTTP/1.1");
    }

    #[test]
    fn request_line_rejects_invalid_separators_and_versions() {
        for line in [
            b"GET  HTTP/1.1".as_slice(),
            b" GET / HTTP/1.1",
            b"GET / HTTP/1.1 ",
            b"GET  / HTTP/1.1",
            b"GET\t/ HTTP/1.1",
            b"GET / HTTP/1.0",
            b"GET / HTTP/1.1\r\n",
        ] {
            assert!(
                RequestLine::parse(line).is_err(),
                "{line:?} should be invalid"
            );
        }
    }
}
